core/src/main/cpp/msg/MessageHeader.cpp

```diff
@@ -167,6 +169,18 @@ MessageHeader::MessageHeader(shared_ptr<MslContext> ctx, shared_ptr<EntityAuthen
     if (!encrypted && headerData->userAuthData)
         throw MslInternalException("User authentication data cannot be included if the message is not encrypted.");
 
+    // Older MSL stacks expect the sender if a master token is being used.
+    //
+    // If the local entity does not know its entity identity, then use the
+    // empty string. This will work except for the case where the old MSL
+    // stack is receiving a message for which it is also the issuer of the
+    // master token. That scenario will continue to fail.
+    shared_ptr<string> sender;
```

Comment:
@padolph FYI this is an example of where using "" as a sentinel value equating to non-existence breaks down. Thus the use of shared_ptr<string> to provide a NULL value if necessary.